spotify-api.el

```diff
@@ -1,55 +1,165 @@
+
+(require 'simple-httpd)
+
+;; Due to an issue related to compilation and the way oauth2 uses defadvice
+;; (including a FIXME as of 0.1.1), this declaration exists to prevent
+;; compiler warnings.
+(eval-when-compile
+  (defvar url-http-method nil)
+  (defvar url-http-data nil)
+  (defvar url-http-extra-headers nil)
+  (defvar oauth--token-data nil)
+  (defvar url-callback-function nil)
+  (defvar url-callback-arguments nil)
+  (require 'oauth2))
+
+(declare-function oauth2-request-access "oauth2")
+
+(defvar *spotify-user*         nil
+  "Cached user object.")
+(defvar *spotify-oauth2-token* nil
+  "Cached OAuth2 token.")
+(defvar *spotify-oauth2-ts*    nil
+  "Unix timestamp in which the OAuth2 token was retrieved.
+This is used to manually refresh the token when it's about to expire.")
 
 (defcustom spotify-oauth2-client-id ""
-  "The unique identifier for your application. More info at
-https://developer.spotify.com/web-api/tutorial/."
+  "The unique identifier for your application.
+More info at https://developer.spotify.com/web-api/tutorial/."
+  :group 'spotify
   :type 'string)
 
 (defcustom spotify-oauth2-client-secret ""
-  "The key that you will need to pass in secure calls to the Spotify Accounts and
-Web API services. More info at
+  "The OAuth2 key provided by Spotify.
+This is the key that you will need to pass in secure calls to the Spotify
+Accounts and Web API services.  More info at
 https://developer.spotify.com/web-api/tutorial/."
+  :group 'spotify
   :type 'string)
 
 (defcustom spotify-api-search-limit 50
   "Number of items returned when searching for something using the Spotify API."
+  :group 'spotify
   :type 'integer)
 
 (defcustom spotify-api-locale "en_US"
-  "Optional. The desired language, consisting of an ISO 639 language code and
-an ISO 3166-1 alpha-2 country code, joined by an underscore.
-For example: es_MX, meaning Spanish (Mexico). Provide this parameter if you
-want the category metadata returned in a particular language."
+  "Optional.  The desired language.
+This consists of an ISO 639 language code and an ISO 3166-1 alpha-2 country
+code, joined by an underscore.  Example: es_MX, meaning Spanish (Mexico).
+Provide this parameter if you want the category metadata returned in a
+ particular language."
+  :group 'spotify
   :type 'string)
 
 (defcustom spotify-api-country "US"
-  "Optional. A country: an ISO 3166-1 alpha-2 country code. Provide this
-parameter if you want to narrow the list of returned categories to those
-relevant to a particular country. If omitted, the returned items will be
+  "Optional.  An ISO 3166-1 alpha-2 country code.
+Provide this parameter if you want to narrow the list of returned categories
+to those to a particular country.  If omitted, the returned items will be
 globally relevant."
+  :group 'spotify
   :type 'string)
 
+(defcustom spotify-oauth2-callback-port "8080"
+  "The port for the httpd to listen on for the OAuth2 callback."
+  :group 'spotify
+  :type 'string)
```

Comment:
Why not sticking with the current port `8591` as the default port? It probably has less chance of conflicting than `8080`.